SofaKernel/modules/SofaBaseMechanics/BarycentricMappers/BarycentricMapperTopologyContainer.inl

```diff
@@ -146,94 +147,107 @@ template <class In, class Out, class MappingDataType, class Element>
 void BarycentricMapperTopologyContainer<In,Out,MappingDataType,Element>::init ( const typename Out::VecCoord& out, const typename In::VecCoord& in )
 {
     initHashing(in);
-
-    const helper::vector<Element>& elements = getElements();
-    helper::vector<Mat3x3d> bases;
-    helper::vector<Vector3> centers;
-
     this->clear ( int(out.size()) );
-    bases.resize ( elements.size() );
-    centers.resize ( elements.size() );
-
-    // Compute bases and centers of each element
-    for ( unsigned int e = 0; e < elements.size(); e++ )
-    {
-        Element element = elements[e];
-
-        Mat3x3d base;
-        computeBase(base,in,element);
-        bases[e] = base;
-
-        Vector3 center;
-        computeCenter(center,in,element);
-        centers[e] = center;
-    }
+    computeBasesAndCenters(in);
 
     // Compute distances to get nearest element and corresponding bary coef
+    const helper::vector<Element>& elements = getElements();
     for ( unsigned int i=0; i<out.size(); i++ )
     {
-        Vec3d outPos = Out::getCPos(out[i]);
-        Vector3 baryCoords;
-        int elementIndex = -1;
-        double distance = std::numeric_limits<double>::max();
+        Vector3 outPos = Out::getCPos(out[i]);
+        NearestParams nearestParams;
 
+        // Search nearest element in grid cell
+        Vec3i gridIds = getGridIndices(outPos);
         unsigned int h = getHashIndexFromCoord(outPos);
-        for ( unsigned int j=0; j<m_hashTable[h].size(); j++)
+        for (unsigned int j=0; j<m_hashTable[h].size(); j++)
         {
-            unsigned int e = m_hashTable[h][j];
-            Vec3d bary = bases[e] * ( outPos - in[elements[e][0]] );
-            double dist;
-            computeDistance(dist, bary);
-            if ( dist>0 )
-                dist = ( outPos-centers[e] ).norm2();
-            if ( dist<distance )
+            HashEntry entry = m_hashTable[h][j];
+            if(sameGridCell(entry, gridIds))
             {
-                baryCoords = bary;
-                distance = dist;
-                elementIndex = int(e);
+                unsigned int e = entry.elementId;
+                Vector3 inPos = in[elements[e][0]];
+                checkDistanceFromElement(e, outPos, inPos, nearestParams);
             }
         }
 
-        if(elementIndex==-1)
+        if(nearestParams.elementIndex==-1) // No element in grid cell, perform exhaustive search
```

Comment:
if you want to be consistent with the hashTable structure and the HashEntry.elementId which is an unsigned int. 
you could put this elementIndex as uint also and use UINT_MAX instead of -1. 
But this is just if you might need to compare elementId and elementIndex.